File: HomeWork/Home5.1/MegaDataPool.java
import java.util.ArrayList;
import java.util.List;

public class MegaDataPool {
    private int poolSize;
    private List<MegaData> pool;
    private boolean[] available;

    public MegaDataPool(int poolSize) {
        this.poolSize = poolSize;
        pool = new ArrayList<>(poolSize);
        available = new boolean[poolSize];
        for (int i = 0; i < poolSize; i++) {
            pool.add(new MegaData());
            available[i] = true;
        }
        System.out.println("MegaDataPool constructor called with pool size: " + poolSize);
    }

    public MegaData acquire() {
        for (int i = 0; i < poolSize; i++) {
            if (available[i]) {
                available[i] = false;
                System.out.println("Acquired object at index: " + i);
                return pool.get(i);
            }
        }
        throw new RuntimeException("No available MegaData objects in the pool");
    }

    public void release(MegaData data) {
        for (int i = 0; i < poolSize; i++) {
            if (pool.get(i) == data) {
                if (!available[i]) {
                    available[i] = true;
                    data.reset();
                    System.out.println("Released object at index: " + i);
                    return;
                } else {
                    System.out.println("Error: This MegaData object is already released");
                    throw new RuntimeException("This MegaData object is already released");
                }
            }
        }
        System.out.println("Error: This MegaData object does not belong to the pool");
        throw new RuntimeException("This MegaData object does not belong to the pool");
    }

    public int size() {
        return poolSize;
    }

    public int usedSize() {
        int count = 0;
        for (boolean available : available) {
            if (!available) {
                count++;
            }
        }
        return count;
    }
}

File: Class/practice5.1/MegaDataPool.cpp
#include "MegaDataPool.h"
#include <iostream>
#include <algorithm> // Include algorithm for std::count

MegaDataPool::MegaDataPool(std::size_t poolSize) : poolSize(poolSize), pool(poolSize), available(poolSize, true) {
    std::cout << "MegaDataPool constructor called with pool size: " << poolSize << "\n";
    std::cout << "Pool and availability vectors initialized.\n";
    std::cout << "MegaDataPool instance address: " << this << std::endl;
}

MegaData& MegaDataPool::acquire() {
    std::cout << "Acquire called\n";
    for (std::size_t i = 0; i < poolSize; ++i) {
        if (available[i]) {
            available[i] = false;
            std::cout << "Acquired object at index: " << i << "\n";
            return pool[i];
        }
    }
    throw std::runtime_error("No available MegaData objects in the pool");
}

void MegaDataPool::release(MegaData& data) {
    std::cout << "Release called\n";
    for (std::size_t i = 0; i < poolSize; ++i) {
        if (&pool[i] == &data) {
            if (!available[i]) {
                available[i] = true;
                data.reset();
                std::cout << "Released object at index: " << i << "\n";
                return;
            } else {
                std::cout << "Error: This MegaData object is already released\n";
                throw std::runtime_error("This MegaData object is already released");
            }
        }
    }
    std::cout << "Error: This MegaData object does not belong to the pool\n";
    throw std::runtime_error("This MegaData object does not belong to the pool");
}

std::size_t MegaDataPool::size() const {
    std::cout << "size() called\n";
    return poolSize;
}

std::size_t MegaDataPool::usedSize() const {
    std::cout << "usedSize() called\n";
    return std::count(available.begin(), available.end(), false);
}
